### ISSUE TITLE
Better documentation for the "mounted" property

### ISSUE THREAD

**Issue by yyoon** [flutter/flutter#6246]
I recently had to implement some async logic with progress indicator, and didn't realize that I should be checking the `mounted` property in the callback, until @apwilson told me about it.

This seems like an important thing to know, but I don't recall seeing this from the tutorials on the official site.

**Comment by Hixie** [comment]
You should in theory not have to use "mounted", so long as you cancel all the work you're doing in your "dispose". "mounted" is more of a lost-ditch check to make sure you don't setState on a dead object.

We should document all this better though, for sure.

**Comment by sethladd** [comment]
Patches welcome for the API docs! :)

**Comment by yyoon** [comment]
@Hixie I'm not sure if `dispose()` would help here. To be more a bit more concrete, here's a simplified version of code I have:

```dart
@override
void initState() {
  super.initState();

  someAsyncFunction().then((result) {
    if (!mounted)
      return;
    setState(() {
      _inProgress = false;
      // set some more states using the result...
    });
  }).catchError((e) {
    if (!mounted)
      return;
    setState(() {
      // set some error states.
    });
  });

  _inProgress = true;
}
```

The async function returns a `Future`, and does some network query internally. I don't think `Future` object has a general way of cancelling its work, which probably means I cannot cancel this work in `dispose()`. Am I missing something here?

**Comment by Hixie** [comment]
The ideal solution is to have some way to cause whatever someAsyncFunction is waiting on to throw with some exception that you recognise here as meaning "the work was canceled" and when you get that you do nothing.

But if you don't mind someAsyncFunction running in the background after it's no longer relevant then checking mounted in that way should work too.

**Comment by yyoon** [comment]
I see. I guess it depends on whether I have the full control over the async logic or not.

May I suggest adding the following statement (taken from the `setState()` dartdoc):

> It is an error to call this method after the framework calls `dispose`. You can determine whether it is legal to call this method by checking whether the `mounted` property is true.

to the Framework Tour page, right after it talks about the `dispose()` method? I can make a PR for that if that makes sense to you.

**Comment by feinstein** [comment]
Also, please provide a snippet on how to throw an exception inside a waiting async call, from `dispose()`.

**Comment by cbracken** [comment]
Related: #6246.

**Comment by felipecastrosales** [comment]
News about this? @Hixie

**Comment by Hixie** [comment]
Documentation has generally improved around this over the years, and we've added lints for it, but it would still be good for someone to take a holistic approach to documenting this more thoroughly, which is why this is still open.
